src/frontend/lexer.rs

```diff
@@ -0,0 +1,259 @@
+use itertools::Itertools;
+use std::iter::Peekable;
+
+use super::token::Token;
+
+// -------------------------------------------------------------------------- //
+// Error type                                                                 //
+// -------------------------------------------------------------------------- //
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Error {
+    UnexpectedSymbol(char),
+    UnclosedString,
+}
+
+// -------------------------------------------------------------------------- //
+// Type aliases                                                               //
+// -------------------------------------------------------------------------- //
+
+type TokResult = Result<Token, Error>;
+
+// -------------------------------------------------------------------------- //
+// Lexer struct                                                               //
+// -------------------------------------------------------------------------- //
+
+pub struct Lexer<I: Iterator<Item = char>> {
+    source: Peekable<I>,
+}
+
+impl<I: Iterator<Item = char>> Lexer<I> {
+    fn new(source: Peekable<I>) -> Self {
+        Self { source }
```

Comment:
This works because of the `From` traits below?

Reply:
Um... No? This just creates an instance of a `Lexer` struct. The only field is of type `Peekable<I>`, which is exactly what gets passed in. `new` is also private. So it's just a very thin helper / wrapper around calling `Self { source: whatever }`